feat: Add trust boundary support for service accounts and impersonation.

nbayati opened the pull request:
design: [go/trust-boundaries-auth-sdk-v2](http://goto.google.com/trust-boundaries-auth-sdk-v2)

nbayati committed:
feat: adds trust boundary lookup support for SA and impersonated credentials

nbayati committed:
Add feature flag, fix ud bug, and update no-op response.

nbayati committed:
Add _build_trust_boundary_lookup_url to external account.

sai-sunder-s commented on google/oauth2/service_account.py:
```diff
@@ -84,13 +84,17 @@
 _DEFAULT_TOKEN_LIFETIME_SECS = 3600  # 1 hour in seconds
 _GOOGLE_OAUTH2_TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
+_TRUST_BOUNDARY_LOOKUP_ENDPOINT = (
+    "https://iamcredentials.{}/v1/projects/-/serviceAccounts/{}/allowedLocations"
```
this url seems to be the same for impersonated cred as well. Will it be same for all supported cred types?

sai-sunder-s commented on tests/test_impersonated_credentials.py:
```diff
@@ -143,6 +148,7 @@ def make_credentials(
         target_principal=TARGET_PRINCIPAL,
         subject=None,
         iam_endpoint_override=None,
+        trust_boundary=NO_OP_TRUST_BOUNDARY,
```
all tests are using NO_OP_TRUST_BOUNDARY? We have to test valid trust boundaries as well.

sai-sunder-s replied:
same for service account tests as well

nbayati commented on google/auth/credentials.py:
```diff
@@ -299,6 +289,124 @@ def with_universe_domain(self, universe_domain):
+        raise NotImplementedError("_build_trust_boundary_lookup_url must be implemented")
+
+    @staticmethod
+    def _parse_trust_boundary(trust_boundary_string: str):
```
Thanks for catching this. It's no longer used so I went ahead and removed it.

nbayati committed:
Implement additional unit tests for the trust boundary

lsirac reviewed (commented):
Can you handle feature gating the same way we do for pluggable auth? We should align this across all languages.

Also, please add the following test cases:

* Please add a test scenario for both service_account.Credentials and impersonated_credentials.Credentials where a valid (non-noop) trust boundary is successfully fetched and cached on the first refresh. A second refresh should then mock a RefreshError from the lookup endpoint. The test should assert that no exception is raised and that the initially cached trust boundary data is preserved on the credentials object. This would validate the fallback logic in _refresh_trust_boundary.

* Include explicit tests for both service_account.Credentials in tests/oauth2/test_service_account.py and impersonated_credentials.Credentials in tests/test_impersonated_credentials.py to confirm that the trust boundary lookup is skipped when the credential is instantiated with a non-default universe_domain. This ensures the check is correctly triggered by the concrete classes.

nbayati committed:
implement trust boundary in compute_engine to support GCE instances.

nbayati committed:
Fix failing unit test, and change acceptable values for the env variable

sai-sunder-s commented on google/auth/compute_engine/credentials.py:
```diff
@@ -111,6 +119,33 @@ def refresh(self, request):
+
+            request = google_auth_requests.Request()
+            try:
+                info = _metadata.get_service_account_info(request, "default")
```
@harkamaljot you had worked on optimizing this call somewhere else in the code recently. Can you PTAL at this method and see if there are any issues in doing this call or something can be optimized.

sai-sunder-s commented on google/auth/compute_engine/credentials.py:
```diff
@@ -179,9 +216,23 @@ def with_universe_domain(self, universe_domain):
+            default_scopes=self._default_scopes,
+            trust_boundary=trust_boundary,
+        )
+        creds._universe_domain = self._universe_domain
```
why not initialize universe domain through the creds init

nbayati replied:
Thanks for the suggestion! That's a great point. I've updated the code to pass universe_domain to the constructor and applied the same pattern to the other with_* methods for consistency.

sai-sunder-s commented on google/auth/credentials.py:
```diff
@@ -299,6 +289,128 @@ def with_universe_domain(self, universe_domain):
+        super().apply(headers, token)
+        if self._trust_boundary is not None:
+            if (
+                self._trust_boundary["encodedLocations"]
+                == NO_OP_TRUST_BOUNDARY_ENCODED_LOCATIONS
```
can use _has_no_op_trust_boundary here?

nbayati committed:
Add unit tests for gce trust boundary.

nbayati committed:
Use no op method instead of directly comparing values

nbayati committed:
fix a typo in calling the method noop

nbayati committed:
Add x-allowed-location header to all IAM requests.

lsirac commented on google/auth/external_account.py:
Lets revert these changes and keep them for a future PR. We should only add this with the changes for supporting the feature

lsirac commented on google/auth/impersonated_credentials.py:
```diff
@@ -251,13 +259,15 @@ def __init__(
     @_helpers.copy_docstring(credentials.Credentials)
     def refresh(self, request):
         self._update_token(request)
+        self._refresh_trust_boundary(request)
```
One suggestion for a potential refactoring to improve the design:

Currently, each class that inherits from `CredentialsWithTrustBoundary` (e.g., `compute_engine.Credentials`, `impersonated_credentials.Credentials`, `service_account.Credentials`) is required to explicitly call `self._refresh_trust_boundary(request)` within its own `refresh` method.

To enhance encapsulation and reduce the burden on subclasses, it would be preferable if this call was handled automatically by the base class. Could this be refactored?

lsirac commented on google/oauth2/_client.py:
```diff
@@ -506,3 +513,138 @@ def refresh_grant(
+        kwargs: Additional arguments passed on to the request method. The
+            kwargs will be passed to `requests.request` method, see:
+            https://docs.python-requests.org/en/latest/api/#requests.request.
+            For example, you can use `cert=("cert_pem_path", "key_pem_path")`
+            to set up client side SSL certificate, and use
+            `verify="ca_bundle_path"` to set up the CA certificates for sever
+            side SSL certificate verification.
```
I don't think we should include this. There isn't really anything that can be added here

lsirac commented on google/oauth2/service_account.py:
How does this work for self signed JWTs?

nbayati replied:
I've updated service_account.Credentials to handle self signed JWTs. When a credential that uses a self-signed JWT also needs to perform a trust boundary lookup, the refresh() method now follows a two-step process:

1. IAM-Specific JWT for Lookup: It first generates a temporary self-signed JWT where the audience is specifically set to the IAM Credentials API. This token is used exclusively to authenticate the request that fetches the trust boundary information.

2. Final JWT for Target API: After the trust boundary is successfully retrieved, the method then proceeds to generate the final self-signed JWT with the audience required by the application's target API (e.g., Pub/Sub, Storage).

This approach ensures that the trust boundary lookup is authenticated correctly using a token the IAM service expects, while the final token used by the application is properly configured for its intended service.

lsirac commented on google/oauth2/_client.py:
```diff
@@ -510,7 +515,7 @@ def refresh_grant(
 
 
-def lookup_trust_boundary(request, url, access_token):
+def lookup_trust_boundary(request, url, headers=None):
```
Shouldn't this be _lookup_trust_boundary?

Also, should this method handle adding auth to the call, instead of that burden being handled by callers?

lsirac reviewed (changes requested):
This is looking good! The main thing is to refactor this, if possible, so that the burden of refreshing trust boundaries is handled by the base class.

For better test coverage, we should add a few test cases for some edge scenarios.
- For the new `with_trust_boundary()` method, it would be great to have tests for each credential type confirming that it returns a new instance and correctly copies all other attributes, modifying only the trust boundary itself.

- What is the expected behavior if `impersonated_credentials` are created with `source_credentials` that don't have trust boundary support? We should probably add a test to ensure the `x-allowed-locations` header is not sent in that case. 

- For `service_account.Credentials` with `always_use_jwt_access=True`, does the trust boundary lookup still work as expected?

- A test case for when `_build_trust_boundary_lookup_url` is called when `service_account_email` is `None` would be valuable. The code seems to handle this by raising a `ValueError`, and it's worth having a test to confirm this behavior.

nbayati committed:
Support self-signed jwt and refactor refresh to handle refreshing trust boundary in the base class.

nbayati committed:
Fix failing unit tests

nbayati committed:
Revert changes to external account class file.

nbayati committed:
Additional unit tests and update some old ones.

nbayati commented:
> * What is the expected behavior if `impersonated_credentials` are created with `source_credentials` that don't have trust boundary support? We should probably add a test to ensure the `x-allowed-locations` header is not sent in that case.
I agree that the header shouldn't be sent in those cases. Added a unit test to cover this scenario. 

> * For `service_account.Credentials` with `always_use_jwt_access=True`, does the trust boundary lookup still work as expected?
Added some logic to support trust boundaries for the self signed jwt flow.

Also implemented the suggested refactoring and test scenarios. 
Will update the design to get clarification on some of the credential types we're not sure about, like the idtoken.

lsirac reviewed (dismissed):
Looks good, great job!

nbayati commented on tests/test_external_account.py:
```diff
@@ -920,7 +920,9 @@ def test_refresh_impersonation_without_client_auth_success(
             "authorization": "Bearer {}".format(token_response["access_token"]),
             "x-goog-api-client": IMPERSONATE_ACCESS_TOKEN_REQUEST_METRICS_HEADER_VALUE,
-            "x-allowed-locations": "0x0",
+            # TODO(negarb): Uncomment and update when trust boundary is supported
```
Thanks for the feedback—I agree it's best to avoid placeholders.

The challenge here is that the commented-out value ("x-allowed-locations": "0x0") is from two years ago and is incorrect under the new design.. Since support for external accounts is out of scope for this particular PR, uncommenting this line would cause tests to fail with the wrong assertion.

I've kept it commented to preserve the test structure for the follow-up PR that will correctly implement this feature for external accounts.

Given this context, would you be okay with keeping these placeholders for this PR, or would you prefer I remove the them and create a separate tracking issue?

nbayati commented on google/auth/credentials.py:
```diff
@@ -299,6 +289,161 @@ def with_universe_domain(self, universe_domain):
+        """
+        raise NotImplementedError("_refresh_token must be implemented")
+
+    def with_trust_boundary(self, trust_boundary):
```
The with_* method is part of a design pattern used for credential objects throughout this library.

We treat credential objects as mostly immutable. You'll see similar factory methods like with_scopes(), with_quota_project(), and with_universe_domain().

sai-sunder-s commented on google/oauth2/service_account.py:
```diff
@@ -451,6 +462,57 @@ def refresh(self, request):
+            # We temporarily set self.token for the base lookup method.
+            # The base lookup method will call self.apply() which adds the
+            # authorization header.
+            original_token = self.token
```
This seems to be welcoming race conditions. What happens if one thread is updating trust boundary and set `self.token` to iam audience and another thread tries to read `self.token`

nbayati replied:
Reverted this change since we're still waiting to hear back from the backend team on whether or not the workaround is necessary. We'll address self signed jwt flow if necessary in a future PR.

sai-sunder-s commented on google/oauth2/service_account.py:
```diff
@@ -806,13 +920,17 @@ def _refresh_with_iam_endpoint(self, request):
         )
         jwt_credentials.refresh(request)
+
+        headers = self._get_trust_boundary_header()
```
you will get NotImplementedError here, thrown from ln 863?

sai-sunder-s replied:
Also, there is no access token in this cred to call trust boundary endpoint

nbayati replied:
I reverted the changes to idtoken as it turns out it doesn't need trust boundaries. Thanks for flagging it!

nbayati commented on google/auth/impersonated_credentials.py:
```diff
@@ -570,6 +602,7 @@ def refresh(self, request):
             "Content-Type": "application/json",
             metrics.API_CLIENT_HEADER: metrics.token_request_id_token_impersonate(),
         }
+        headers.update(self._target_credentials._get_trust_boundary_header())
```
You're right. Thank you for catching this.

nbayati committed:
Revert changes to idtoken

nbayati committed:
Revert the self signed jwt token workaround

nbayati committed:
revert idtoken and jwt trust boundary tests.

nbayati committed:
Fix failing github check

nbayati committed:
Merge branch 'main' into feat/trust-boundary-service-account

nbayati committed:
Fix failing unit tests in compute engine

nbayati committed:
fix linter issues

nbayati committed:
Fix linter issues

nbayati committed:
remove trust boundary related code from idtoken as it is not a supported credential type

nbayati committed:
remove unused line in test

nbayati committed:
Merge branch 'main' into feat/trust-boundary-service-account
